### PR TITLE
Ensemble reduced functional

### PR DESCRIPTION
**Pull request by Ig-dolci**

### CODE REVIEW

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+from pyadjoint import ReducedFunctional
+from pyop2.mpi import MPI
+
+import firedrake
+
+
+class EnsembleReducedFunctional(ReducedFunctional):
+    """Enable solving simultaneously reduced functionals in parallel.
+
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:9]: Add the math of ensemble reduced functional to make clear what is going on.

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        if isinstance(local_functional, float):
+            total_functional = self.ensemble.ensemble_comm.allreduce(sendobj=local_functional, op=MPI.SUM)
+        elif isinstance(local_functional, firedrake.Function):
+            total_functional = firedrake.Function(local_functional.function_space())
+            total_functional = self.ensemble.allreduce(local_functional, total_functional)
+        else:
+            raise NotImplementedError("This type of functional is not supported.")
+        return total_functional
+
+    def derivative(self, adj_input=1.0, options=None, op=MPI.SUM):
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:45]: remove `op=MPI.SUM`

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    ----------
+    J : pyadjoint.OverloadedType
+        An instance of an OverloadedType, usually :class:`pyadjoint.AdjFloat`.
+        This should be the functional that we want to reduce.
+    control : pyadjoint.Control or list of pyadjoint.Control
+        A single or a list of Control instances, which you want to map to the functional.
+    ensemble : Ensemble
+        An instance of the :class:`~.ensemble.Ensemble`. It is used to communicate the
+        functionals and their derivatives between the ensemble members.
+
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:20]: We should document that J is defined over a `self.ensemble.comm`.

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    See Also
+    --------
+    :class:`~.ensemble.Ensemble`, :class:`pyadjoint.ReducedFunctional`.
+
+    Notes
+    -----
+    To understand more about how ensemble parallelism works, please refer to the Firedrake
+    `documentation <https://www.firedrakeproject.org/parallelism.html#id8>`_.
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:28]: `Firedrake manual <https://www.firedrakeproject.org/parallelism.html#id8>`_.

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    def __init__(self, J, control, ensemble):
+        super(EnsembleReducedFunctional, self).__init__(J, control)
+        self.ensemble = ensemble
+
+    def __call__(self, values):
+        local_functional = super(EnsembleReducedFunctional, self).__call__(values)
+        if isinstance(local_functional, float):
+            total_functional = self.ensemble.ensemble_comm.allreduce(sendobj=local_functional, op=MPI.SUM)
+        elif isinstance(local_functional, firedrake.Function):
+            total_functional = firedrake.Function(local_functional.function_space())
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:39]:
```suggestion
            total_functional = type(local_functional)(local_functional.function_space())
```

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+        See Also
+        --------
+        :meth:`~.ensemble.Ensemble.allreduce`, :meth:`pyadjoint.ReducedFunctional.derivative`.
+        """
+        dJdm_local = super(EnsembleReducedFunctional, self).derivative(adj_input=adj_input, options=options)
+        if isinstance(dJdm_local, list):
+            dJdm_total = []
+            for dJdm in dJdm_local:
+                dJdm_total.append(
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:71]: See `Enlist` in pyadjoint and improve this code.

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+                    self.ensemble.allreduce(dJdm, firedrake.Function(dJdm.function_space()), op=op)
+                )
+        elif isinstance(dJdm_local, firedrake.Function):
+            dJdm_total = firedrake.Function(dJdm_local.function_space())
+            dJdm_total = self.ensemble.allreduce(dJdm_local, dJdm_total, op=op)
+        else:
```
Comment (Ig-dolci) [inline firedrake/adjoint/ensemble_reduced_functional.py:77]: Add the case where the local derivative can be a float.

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        functionals and their derivatives between the ensemble members.
+
+    See Also
+    --------
+    :class:`~.ensemble.Ensemble`, :class:`pyadjoint.ReducedFunctional`.
+
+    Notes
+    -----
+    The functionals :math:`J_i` and the control must be defined over a common
+    `self.ensemble.comm` communicator. To understand more about how ensemble parallelism
```
Comment (dham) [inline firedrake/adjoint/ensemble_reduced_functional.py:49]:
```suggestion
    `ensemble.comm` communicator. To understand more about how ensemble parallelism
```

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        """Compute derivatives of a functional with respect to the control parameters.
+
+        Parameters
+        ----------
+        adj_input : float
+            The adjoint input.
+        options : dict
+            Additional options for the derivative computation.
+        op : mpi4py.MPI.Op
+            The employed MPI operation for the `Ensemble.allreduce` the derivatives.
```
Comment (dham) [inline firedrake/adjoint/ensemble_reduced_functional.py:78]:
```suggestion
```

File: firedrake/adjoint/ensemble_reduced_functional.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+        Returns
+        -------
+            dJdm_total : :class:`~.function.Function` or list of :class:`~.function.Function`
+            The result of Allreduce operations of ``dJdm_local`` into ``dJdm_total`` over the`Ensemble.ensemble_comm`.
```
Comment (dham) [inline firedrake/adjoint/ensemble_reduced_functional.py:83]: should be overloaded type rather than Function.

**Review by dham — DISMISSED** [review]

**Comment by colinjcotter** [comment]
This looks interesting. What’s the motivating application for it?

**Comment by Ig-dolci** [comment]
> This looks interesting. What’s the motivating application for it?

Ensemble functional will be helpful when your cost function `J` is a sum of multiple functionals `J_i`, with each `J_i`  requiring the computation of an adjoint-based gradient. That is typically seen in full waveform inversion, a technique used to create subsurface imaging.

**Review by connorjward — CHANGES_REQUESTED** [review]
I think the tests should run on more ranks to stress the spatial parallelism. Otherwise very happy to approve.

**Review by connorjward — APPROVED** [review]

**Pull request merged**